ingest/expression_writer.py

```diff
@@ -0,0 +1,268 @@
+"""expression_writer.py
+extract gene-level expression data from dense/sparse matrix files and process in the
+context of a cluster file's cells.
+this mimics the `expression` data array in visualization API responses
+
+EXAMPLES (must be invoked via ingest_pipeline.py)
+
+dense matrix:
+python3 ingest_pipeline.py --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 \
+                             render_expression_arrays --matrix-file-path ../tests/data/dense_expression_matrix.txt \
+                             --matrix-file-type dense \
+                             --cluster-file ../tests/data/cluster_example.txt \
+                             --cluster-name 'Dense Example' --render-expression-arrays
+
+sparse matrix:
+python3 ingest_pipeline.py --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 \
+                             render_expression_arrays --matrix-file-path ../tests/data/mtx/matrix_with_header.mtx \
+                             --matrix-file-type mtx \
+                             --gene-file ../tests/data/mtx/sampled_genes.tsv \
+                             --barcode-file ../tests/data/mtx/barcodes.tsv \
+                             --cluster-file ../tests/data/mtx/cluster_mtx_barcodes.tsv \
+                             --cluster-name 'Sparse Example' --render-expression-arrays
+"""
+from __future__ import annotations
+import os
+import subprocess
+import re
+import multiprocessing
+import sys
+import datetime
+from dateutil.relativedelta import relativedelta
+from functools import partial
+
+try:
+    from writer_functions import round_exp, encode_cluster_name, open_file, make_data_dir, get_matrix_size, \
+        get_cluster_cells, load_entities_as_list, process_sparse_fragment, write_gene_scores, process_dense_line, \
+        COMMA_OR_TAB
+    from monitor import setup_logger
+    from ingest_files import IngestFiles
+except ImportError:
+    from .writer_functions import round_exp, encode_cluster_name, open_file, make_data_dir, get_matrix_size, \
+        get_cluster_cells, load_entities_as_list, process_sparse_fragment, write_gene_scores, process_dense_line, \
+        COMMA_OR_TAB
+    from .monitor import setup_logger
+    from .ingest_files import IngestFiles
```

Comment:
I introduced this inelegant import technique to deal with a quirk of supporting CSFV in our public CLI / `manage_study.py`.  We no longer support the public CLI, so it'd make sense to remove these `try`/ `except` clauses in general.  And (in particular) we wouldn't call this module from the public CLI, so we can certainly omit it here.

```suggestion
from writer_functions import round_exp, encode_cluster_name, open_file, make_data_dir, get_matrix_size, \
    get_cluster_cells, load_entities_as_list, process_sparse_fragment, write_gene_scores, process_dense_line, \
    COMMA_OR_TAB
from monitor import setup_logger
from ingest_files import IngestFiles
```